crates/turborepo-lib/src/globwatcher/mod.rs

```diff
@@ -0,0 +1,206 @@
+                .cloned()
+                .collect::<Vec<_>>()
+        };
+
+        let mut stream = match self.watcher.lock().expect("no panic").take() {
+            Some(watcher) => watcher.into_stream(token),
+            None => {
+                warn!("watcher already consumed");
+                return;
+            }
+        };
+
+        // watch all the globs currently in the map
+        for glob in start_globs {
+            self.config.include(glob.to_string()).await.unwrap();
+        }
+
+        while let Some(Ok(event)) = stream.next().await {
+            trace!("event: {:?}", event);
+
+            let repo_relative_paths = event
+                .paths
+                .iter()
+                .filter_map(|path| path.strip_prefix(&root_folder).ok());
+
+            // put these in a block so we can drop the locks before we await
+            let globs_to_exclude = {
+                let glob_statuses = self.glob_statuses.lock().expect("ok");
+                let hash_globs = self.hash_globs.lock().expect("ok");
+
+                // hash globs is unlocked after this
+                let (hash_globs_to_clear, globs_to_exclude) =
+                    populate_hash_globs(&glob_statuses, repo_relative_paths, hash_globs);
+
+                clear_hash_globs(glob_statuses, hash_globs_to_clear);
+
+                globs_to_exclude
+            };
+
+            for glob in globs_to_exclude {
+                self.config.exclude(glob.to_string()).await.unwrap();
+            }
+        }
+    }
+
+    pub async fn watch_globs<Iter: IntoIterator<Item = String>>(
+        &self,
+        hash: String,
+        include: Iter,
+        exclude: Iter,
+    ) {
+        self.config.flush().await.unwrap();
+
+        let hash = Arc::new(hash);
+        let include: HashSet<_> = include.into_iter().map(Arc::new).collect();
+        let exclude = exclude.into_iter().map(Arc::new).collect();
+
+        for glob in include.iter() {
+            self.config.include(glob.to_string()).await.unwrap();
+        }
+
+        {
+            let mut glob_status = self.glob_statuses.lock().expect("no panic");
+            glob_status
+                .entry(hash.clone())
+                .or_default()
+                .extend(include.clone());
+        }
+
+        {
+            let mut hash_globs = self.hash_globs.lock().expect("no panic");
+            hash_globs.insert(hash, Glob { include, exclude });
+        }
+    }
+
+    /// Given a hash and a set of candidates, return the subset of candidates
+    /// that have changed.
+    pub async fn changed_globs(
+        &self,
+        hash: &str,
+        mut candidates: HashSet<String>,
+    ) -> HashSet<String> {
+        self.config.flush().await.unwrap();
+
+        let globs = self.hash_globs.lock().unwrap();
+        match globs.get(&Arc::new(hash.to_string())) {
+            Some(glob) => {
+                candidates.retain(|c| glob.include.contains(c));
```

Comment:
Is this backwards? If there's some new glob we don't know about in `candidates` it will get pruned here, but we don't actually know if anything changed because we (maybe) weren't watching it. 

IIUC, `hash_globs` contains the globs that _haven't_ changed.

Reply:
Yes, you are correct. We want to retain the items in candidate that are _not_ in glob.